dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -276,7 +276,7 @@ private ListenableFuture<Response> handleThrowable(Throwable clientSideThrowable
                 if (requestCanBeRetried() && shouldAttemptToRetry(clientSideThrowable)) {
                     callsiteStacktrace.ifPresent(clientSideThrowable::addSuppressed);
                     Meter retryReason = retryDueToThrowable.apply(clientSideThrowable);
-                    long backoffNanoseconds = getBackoffNanoseconds();
+                    long backoffNanoseconds = failures <= 1 ? 0 : getBackoffNanoseconds();
```

Comment:
@fsamuel-bs I know I arrive after the battle and this has already merged (I was ooto), but would it be possible to inline some comment/explanation as to 
- why we're doing this
- why we're only doing it here and not on the `handleHttpResponse` code-path
?

As is, even reading the PR description, it's not entirely obvious, so a quick comment might go a long way

Reply:
We should definitely update the PR description, but responding to the specific points:

It's relatively common for requests to fail when a node that is not live, often because it is restarting. In these cases we want to quickly try again on another node. The request didn't fail because the server didn't have capacity, etc. so there's no reason to wait before trying again.

We don't want to do this on the `handleHttpResponse` code path because we received a response from the server, so we know that node is at least live.

Reply:
> > If a request failed, the call is immediately retried on a second host.
>
> To be clear, there is no guarantee that the retried will requests will be routed to a different node.
>
> But in practice we expect that a second attempt will typically succeed because the IOException from the first attempt [updates the balanced score tracker](https://github.com/palantir/dialogue/blob/1f21912856b571a71a3a896e61f2a65efec760d5/dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedScoreTracker.java#L191) in a way that makes it much less likely that the original node will be selected for the second attempt.

I do think it would be prudent for us to try an validation this implicit assumption being made here. If this assumption is not true, then this change may more more harmful than beneficial.

Reply:
The rationale makes sense to me, but I think it should be documented in the code, e.g.
```suggestion
                    // Retry immediately when we failed to reach the server, since it wasn't a capacity issue
                    // Note that this should retry on another node
                    long backoffNanoseconds = failures <= 1 ? 0 : getBackoffNanoseconds();
```
I'm not sure what validation we could make to verify we didn't reach the server though 🤔 (not familiar enough with the types of throwables we can get here)

Reply:
https://github.com/palantir/dialogue/pull/2657